Add addition info for Kysely (#107)
xataio/mdx-docs: 070-Integrations/040-kysley.mdx

---
title: Kysley with Xata
navTitle: Kysley
keywords: ['orm', 'kysley', 'typescript', 'query builder']
description: "Use the Kysley integration as a TypeScript SQL query builder with Xata"
slug: integrations/kysley
published: true
---

[Kysely](https://kysely.dev/docs/intro) is a TypeScript SQL query builder. It focuses on ensuring type safety and autocompletion and works with a growing number of database services. Kysely works in any JavaScript runtime, including Node.js, Cloudflare Workers, and Deno.

While the native Xata TypeScript SDK offers similar built-in functionality, using Kysely might be preferable if you are migrating from another database service, or if you are already using Kysely in your project.

This integration consists of a [Xata dialect for Kysely](https://github.com/xataio/client-ts/tree/main/packages/plugin-client-kysely).

### Installation

Install both Kysely and `@xata.io/kysely` with the Xata client. You can use the following package managers:

<TabbedCode tabs={['npm', 'pnpm', 'yarn']}>

> 
> ```sh
> npm install kysely @xata.io/kysely @xata.io/client
> ```
> 
> ```sh
> pnpm add kysely @xata.io/kysely @xata.io/client
> ```
> 
> ```sh
> yarn add kysely @xata.io/kysely @xata.io/client
> ```
> 

Reviewer: ````suggestion
```sh
npm install kysely @xata.io/kysely @xata.io/client
```

```sh
pnpm add kysely @xata.io/kysely @xata.io/client
```

```sh
yarn add kysely @xata.io/kysely @xata.io/client
```
````

---
title: Kysley with Xata
navTitle: Kysley

> keywords: ['orm', 'kysley']

Reviewer: Kysely is a query builder, drizzle is an ORM

description: "Using the Kysley TypeScript ORM with Xata"
slug: integrations/kysley
published: true
---

[Kysely](https://kysely.dev/docs/intro) is a TypeScript SQL query builder. It focuses on ensuring type safety and autocompletion and works with a growing number of database services. Kysely works in any JavaScript runtime, including Node.js, Cloudflare Workers, and Deno.

While the native Xata TypeScript SDK offers similar built-in functionality, using Kysely might be preferable if you are migrating from another database service, or if you are already using Kysely in your project.

This integration consists of a [Xata dialect for Kysely](https://github.com/xataio/client-ts/tree/main/packages/plugin-client-kysely).

### Installation

Install both Kysely and `@xata.io/kysely` with the Xata client. You can use the following package managers:

<TabbedCode tabs={['npm', 'pnpm', 'yarn']}>

> ```sh
> pnpm add kysely @xata.io/kysely @xata.io/client
> ```
> 
> ```sh
> yarn add kysely @xata.io/kysely @xata.io/client
> ```
> 
> ```sh
> npm install kysely @xata.io/kysely @xata.io/client
> ```

Reviewer: ````suggestion
```sh
npm install kysely @xata.io/kysely @xata.io/client
```

```sh
pnpm add kysely @xata.io/kysely @xata.io/client
```

```sh
yarn add kysely @xata.io/kysely @xata.io/client
```
````

</TabbedCode>

### Usage 

Kysely is a query builder library for TypeScript that helps in generating SQL queries. For compatibility with Xata, a specialized dialect has been developed for Kysely. The [Xata dialect for Kysely](https://github.com/xataio/client-ts/tree/main/packages/plugin-client-kysely) enables the use of Xata-specific functionality with Kysely.

The following example, demonstrates how you can employ the Xata dialect for Kysely:

```tsx
import { Kysely } from 'kysely';
import { XataDialect, Model } from '@xata.io/kysely';

> import { DatabaseSchema, getXataClient } from 'xata';   // generated client

Reviewer: ```suggestion
import { DatabaseSchema, getXataClient } from 'xata'; // Generated client
```

const xata = getXataClient();

const db = new Kysely<Model<DatabaseSchema>>({
  dialect: new XataDialect({ xata }),
});

const drivers = await db
  .selectFrom("drivers")
  .select(["forename", "surname"])
  .where("nationality", "=", "Spanish")
  .execute();

console.log(drivers);
```

In the provided code example, the following steps are being taken to integrate Xata with Kysley:

1. Install the Xata client as usual, see our [getting started guide](/docs/getting-started/installation).

> 1. Import `getXataClient` from the generated client (usually generated by `xata pull` in `src/xata.ts`).

Reviewer: ```suggestion
1. Import `getXataClient` from the generated client (usually generated by `xata pull` in a file of your project).
```

> 1. Instantiate Kysley with the Xata dialect. Note that the `Model` type adapts the Xata generated `DatabaseSchema` type to the `Model` type used by Kysely.

Reviewer: ```suggestion
1. Instantiate Kysley with the Xata dialect. Note that the `Model` type adapts the Xata generated `DatabaseSchema` type to a `Model` compatible with Kysely.
```

> 1. Use Kysley as you would normally do. Type safety and autocompletion should work.

Reviewer: ```suggestion
1. Use Kysley as you would normally do with type safety and autocompletion.

```
